osu.Server.Queues.ScoreStatisticsProcessor/Commands/Maintenance/MigratePlaylistScoresToSoloScoresCommand.cs

```diff
@@ -0,0 +1,199 @@
+                            List<HitResult> allHits = statistics
+                                                      .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value))
+                                                      .ToList();
+                            var maximumStatistics = new Dictionary<HitResult, int>();
+
+                            foreach (var groupedStats in allHits
+                                                         .Select(r => getMaxJudgementFor(r, maxRulesetJudgement))
+                                                         .GroupBy(r => r))
+                            {
+                                maximumStatistics[groupedStats.Key] = groupedStats.Count();
+                            }
+
+                            APIMod[] roomMods = JsonConvert.DeserializeObject<APIMod[]>(item.required_mods)!;
+                            APIMod[] scoreMods = JsonConvert.DeserializeObject<APIMod[]>(score.mods)!;
+
+                            foreach (var m in roomMods)
+                                Debug.Assert(scoreMods.Contains(m));
+
+                            int? insertId = null;
+
+                            if (score.total_score != null)
+                            {
+                                var soloScoreInfo = SoloScoreInfo.ForSubmission(new ScoreInfo
+                                {
+                                    Statistics = statistics,
+                                    MaximumStatistics = maximumStatistics,
+                                    Ruleset = ruleset.RulesetInfo,
+                                    MaxCombo = (int)score.max_combo!,
+                                    APIMods = scoreMods,
+                                    Accuracy = (double)score.accuracy!,
+                                    TotalScore = (long)score.total_score,
+                                    Rank = Enum.Parse<ScoreRank>(score.rank),
+                                    Passed = score.passed,
+                                });
+
+                                soloScoreInfo.StartedAt = score.started_at;
+                                soloScoreInfo.EndedAt = score.started_at;
+                                soloScoreInfo.BeatmapID = (int)score.beatmapId;
+
+                                insertId = await db.InsertAsync(new SoloScore
+                                {
+                                    user_id = (int)score.userId,
+                                    beatmap_id = (int)score.beatmapId,
+                                    ruleset_id = item.ruleset_id,
+                                    preserve = true,
+                                    ScoreInfo = soloScoreInfo,
+                                    created_at = (DateTimeOffset)score.created_at!,
+                                    updated_at = (DateTimeOffset)score.updated_at!,
```

Comment:
So this is another potential minefield... If the machine that this code is going to be running on will be running on the UTC timezone, this is gonna _work_. But if it's not, then this is gonna silently break.

As far as I can tell, `multiplayer_scores` is keeping dates in `created_at` and `updated_at` as UTC. _However_, because computers are cursed, `DateTime` instances returned by dapper have `DateTimeKind.Unspecified` set. Now the `DateTimeOffset` conversion [assumes that `Unspecified` dates are local](https://github.com/dotnet/runtime/blob/1487d46e4ff7731b52af92b1228373a2a66ed6bf/src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.cs#L85-L89), mostly because it _has_ to flip a coin in such a case. But what that means, is that _if_ the machine that this is ran on doesn't use UTC timezone, then the datetimes stored to db will be _skewed back_ by the inverse of the local computer's timezone offset.

I _believe_ that this is a concern, then the more correct way to convert would be

```suggestion
                                    created_at = new DateTime(score.created_at!.Value.Ticks, DateTimeKind.Utc),
                                    updated_at = new DateTime(score.updated_at!.Value.Ticks, DateTimeKind.Utc),
```

But this _may_ be irrelevant and it's probably best that you double-check this, because this crap is _deeply_ confusing and screwed.

The same applies to `soloScoreInfo.{Started,Ended}At` above.